### PR TITLE
Add directories to cache in one pass

### PR DESCRIPTION
**Pull request by BanzaiMan**
To avoid scanning the same archive multiple times to extract
directories, we pass directories to 'casher add' in one go.

### CODE REVIEW

File: lib/travis/build/script/shared/directory_cache/s3.rb [hunk 1]
```diff
@@ -77,8 +77,8 @@ def install
             end
           end
 
-          def add(path)
-            run('add', path) if path
+          def add(*paths)
+            run('add', paths) if paths
```
Comment (meatballhat) [inline lib/travis/build/script/shared/directory_cache/s3.rb]: What about command line arg limit?  Can you do batches of 200?

File: lib/travis/build/script/shared/directory_cache/s3.rb [hunk 2]
```diff
@@ -77,8 +81,10 @@ def install
             end
           end
 
-          def add(path)
-            run('add', path) if path
+          def add(*paths)
+            if paths
+              paths.flatten.each_slice(ADD_DIR_MAX) { |dirs| run('add', dirs) }
```
Comment (meatballhat) [inline lib/travis/build/script/shared/directory_cache/s3.rb]: :heart_eyes_cat:

**Comment by BanzaiMan** [comment]
Casher change has been deployed to production. I plan to deploy this early tomorrow.

I'm a bit ambivalent about documenting this behavior; on one hand, it is good to speed up extraction via `cache.directories`. On the other hand, this makes other caching niceties (such as `cache: bundler`) less attractive if it is used in conjunction with `cache.directories`. (It is pure win, but it discourages `cache: bundler` to some extent.)

**Comment by BanzaiMan** [comment]
This was merged with https://github.com/travis-ci/travis-build/commit/c09d12807c2f003cebebd5aebb9f8b0a84305552.

**Pull request closed without merging**